File: apps/mobile/consumer-ios/Sources/ConsumerIOSShell/ConsumerAuthSessionManager.swift
import Foundation

public enum ConsumerAuthSessionManagerError: Error {
    case missingSession
}

public struct ConsumerAuthSessionManager {
    private let client: ConsumerBackendClient
    private var activeSession: SessionExchangeResponse?

    public init(client: ConsumerBackendClient) {
        self.client = client
    }

    public mutating func signIn(
        oidcToken: String,
        traceId: String,
        deviceId: String
    ) async throws -> SessionExchangeResponse {
        let response = try await client.exchangeSession(
            oidcToken: oidcToken,
            traceId: traceId,
            deviceId: deviceId
        )
        activeSession = response
        return response
    }

    public mutating func refresh(
        traceId: String,
        deviceId: String
    ) async throws -> SessionExchangeResponse {
        guard let refreshToken = activeSession?.tokenPair.refreshToken else {
            throw ConsumerAuthSessionManagerError.missingSession
        }

        let response = try await client.refreshSession(
            refreshToken: refreshToken,
            traceId: traceId,
            deviceId: deviceId
        )
        activeSession = response
        return response
    }

    public func currentSession() -> SessionExchangeResponse? {
        activeSession
    }
}

File: apps/mobile/consumer-android/src/main/kotlin/com/fleetfeast/mobile/consumer/ConsumerAuthSessionManager.kt
package com.fleetfeast.mobile.consumer

class ConsumerAuthSessionManager(
  private val client: ConsumerBackendClient,
) {
  private var activeSession: SessionExchangeResponse? = null

  fun signIn(
    oidcToken: String,
    traceId: String,
    deviceId: String,
  ): SessionExchangeResponse {
    val response = client.exchangeSession(
      oidcToken = oidcToken,
      traceId = traceId,
      deviceId = deviceId,
    )
    activeSession = response
    return response
  }

  fun refresh(
    traceId: String,
    deviceId: String,
  ): SessionExchangeResponse {
    val refreshToken = activeSession?.tokenPair?.refreshToken
      ?: throw IllegalStateException("No active session to refresh")

    val response = client.refreshSession(
      refreshToken = refreshToken,
      traceId = traceId,
      deviceId = deviceId,
    )
    activeSession = response
    return response
  }

  fun currentSession(): SessionExchangeResponse? {
    return activeSession
  }
}
